onnxruntime/core/providers/cpu/tensor/expand.cc

```diff
@@ -0,0 +1,157 @@
+  const auto* input_data = input_tensor->template Data<T>();
+  const auto& input_shape = input_tensor->Shape().GetDims();
+
+  const auto* input_dims = input_shape.data();
+  auto input_dims_size = static_cast<int64_t>(input_shape.size());
+
+  const auto* shape_tensor = context->Input<Tensor>(1);
+  const auto* shape_dims = shape_tensor->Data<int64_t>();
+  std::vector<int64_t> output_shape{shape_dims, shape_dims + shape_tensor->Shape().Size()};
+
+  if (input_shape.size() > output_shape.size()) {
+    output_shape.insert(output_shape.begin(), input_shape.size() - output_shape.size(), 1);
+  }
+
+  auto input_shape_iter = input_shape.rbegin();
+  auto output_shape_iter = output_shape.rbegin();
+  while (input_shape_iter != input_shape.rend() &&
+         output_shape_iter != output_shape.rend()) {
+    if (*input_shape_iter != *output_shape_iter) {
+      if (1 == *output_shape_iter) {
+        *output_shape_iter = *input_shape_iter;
+      } else if (1 != *input_shape_iter) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "invalid expand shape");
+      }
+    }
+    input_shape_iter++;
+    output_shape_iter++;
+  }
+
+  TensorShape output_tensor_shape(output_shape);
+  auto* output_tensor = context->Output(0, output_tensor_shape);
+  auto* output_data = output_tensor->template MutableData<T>();
+  auto* output_dims = output_shape.data();
+  auto output_dims_size = static_cast<int64_t>(output_shape.size());
+  auto max_dims_size = std::max(input_dims_size, output_dims_size);
+  std::unique_ptr<int64_t[]> input_dim_group{new int64_t[max_dims_size]};
+  std::unique_ptr<int64_t[]> output_dim_group{new int64_t[max_dims_size]};
+  std::unique_ptr<int64_t[]> expand_dim_size{new int64_t[max_dims_size]};
+  auto dim_group_start = max_dims_size;
+
+  for (int64_t input_dims_iter = input_dims_size - 1,
+               output_dims_iter = output_dims_size - 1,
+               last_dim_size = 1,
+               input_count = 1,
+               output_count = 1;
+       output_dims_iter > -1;
+       input_dims_iter--, output_dims_iter--) {
+    auto input_dim = input_dims_iter > -1 ? input_dims[input_dims_iter] : 1;
+    auto output_dim = output_dims[output_dims_iter];
+
+    input_count *= input_dim;
+    output_count *= output_dim;
+
+    if (input_dim == 1 && output_dim > 1 || output_dims_iter == 0) {
+      --dim_group_start;
+      input_dim_group[dim_group_start] = input_count;
+      output_dim_group[dim_group_start] = output_count;
+      expand_dim_size[dim_group_start] = output_count / input_count / last_dim_size;
+      last_dim_size *= expand_dim_size[dim_group_start];
+    }
+  }
+
+  auto distribute_count = input_dim_group[dim_group_start] / input_dim_group[max_dims_size - 1];
+  std::vector<int64_t> output_offsets(distribute_count, 0);
+  int64_t copy_len = input_dim_group[max_dims_size - 1];
+  auto copy_byte = copy_len * sizeof(T);
+
+  concurrency::ThreadPool::TrySimpleParallelFor(
```

Comment:
Why we have two different call `concurrency::ThreadPool::TrySimpleParallelFor`? Is it possible to merge them?

Reply:
No, they are doing totally different things - the first is to distribute data from input to output at right offset, while the second is purely expanding.